[3.x] Add PCF25 shadow filtering option in the GLES3 renderer

Calinou opened the pull request:
- 3.x version of https://github.com/godotengine/godot/pull/92090.

This is a high-quality shadow filter which results in smooth, stable shadows even for dynamic objects.

Thanks to early bailing, PCF25 is faster than the old PCF13 implementation while looking better. It's also not that much slower compared to the new PCF13 implementation.

## Preview

| Disabled | PCF5 | PCF13 | PCF25 |
|-|-|-|-|
| ![2021-10-28_16 55 12](https://user-images.githubusercontent.com/180032/139285142-116e43dc-cd46-47f6-9a52-67c1f4bed8b9.png) | ![2021-10-28_16 54 44](https://user-images.githubusercontent.com/180032/139285137-6aaa8122-9e84-4e40-93a0-d70151761064.png) | ![2021-10-28_16 57 18](https://user-images.githubusercontent.com/180032/139285147-6ade1267-ab76-4de5-9820-d103b2f2aa73.png) | ![2021-10-28_16 55 42](https://user-images.githubusercontent.com/180032/139285145-6b8a9539-62cc-4948-9aa6-a2bf711ecc7a.png) |

### PCF13 with 2048 shadow size (faster but less smooth)

https://user-images.githubusercontent.com/180032/139285063-1de5a7f3-4241-4796-a809-73664e681f67.mp4

### PCF25 with 4096 shadow size (slower but smoother)

https://user-images.githubusercontent.com/180032/139285079-a0b6a65b-2c5f-42bd-8b2c-c79249da4a19.mp4

Calinou committed:
Add PCF25 shadow filtering option in the GLES3 renderer

This is a high-quality shadow filter which results in smooth,
stable shadows even for dynamic objects.

Thanks to early bailing, PCF25 is faster than the old PCF13
implementation while looking better. It's also not that much slower
compared to the new PCF13 implementation.

lawnjelly commented:
Bit of a bodge, but as this is an enum rather than needing conditionals for each (i.e. you are either using 0, 5, 13 or 25) maybe you can get by with two conditional flags...

00 = 0
01 = 5
10 = 13
11 = 25

But yeah sounds like it could do with some looking at the conditional code to see if there's a better solution longterm.

Calinou commented:
Rebased and tested again, it works as expected.

I've changed the early bailing implementation to also take the center sample into account. This is a tad slower as it requires one more sample, but it should lead to less artifacts, particularly with PCF25. This matches the implementation used in the 4.x Compatibility rendering method.

lawnjelly commented on drivers/gles3/shaders/scene.glsl:
```diff
@@ -1270,17 +1270,62 @@ LIGHT_SHADER_CODE
-#ifdef SHADOW_MODE_PCF_13 //ubershader-runtime
+	float avg = textureProj(shadow, vec4(pos, depth, 1.0));
+#ifdef SHADOW_MODE_PCF_LOW //ubershader-runtime
+#ifdef SHADOW_MODE_PCF_HIGH //ubershader-runtime
```
Can we do this? :thinking: 
```
#if defined(SHADOW_MODE_PCF_LOW) && defined(SHADOW_MODE_PCF_LOW)
```
This pattern seems to be used elsewhere in the shaders and it seems easier to follow.

Calinou replied:
I think this messes with `//ubershader-runtime`, unfortunately. We can use this approach in 4.x though where ubershaders aren't used in Compatibility.

lawnjelly commented:
I tried this out and it seems to work fine.

Not really for this PR, but it strikes me at 25 samples, we should be making the PCF radius adjustable per light. With a uniform multiplier in `sample_shadow()` for `shadow_pixel_size` we could adjust the softness, and trade off with the PCF samples.

Also I do wonder for PCF 25 whether a regular pattern gives the best results versus random (even fixed random pattern). Again maybe this should be adjustable. Hopefully we can greatly improve shadows in 3.7.

Calinou commented:
> Not really for this PR, but it strikes me at 25 samples, we should be making the PCF radius adjustable per light. With a uniform multiplier in sample_shadow() for shadow_pixel_size we could adjust the softness, and trade off with the PCF samples.

There's an ~~app~~ PR for it :slightly_smiling_face: https://github.com/godotengine/godot/pull/60243

lawnjelly commented:
> There's an app PR for it 🙂 https://github.com/godotengine/godot/pull/60243

Great! I'm keen for us to improve the shadows in 3.7, including e.g. fixed filter per light or PCSS or similar, and normal bias offset. We'll probably discuss these PRs in more detail over coming months and decide which approaches to go for.
